### PR TITLE
[FIX](Map)fix calculate map offset in olap convertor

### PR DESCRIPTION
**Pull request by amorynan**
# Proposed changes

## Problem summary
When I insert data with bigger kv pair in one row for map , here  coming a core :

```
WRITE of size 16 at 0x625055819400 thread T556 (BaseCompactionT)
    #0 0x564829a56df3 in _mm_store_si128(long long __vector(2)*, long long __vector(2)) /mnt/disk1/wangqiannan/tool/ldb_toolchain/lib/gcc/x86_64-linux-gnu/11/include/emmintrin.h:734
    #1 0x564829a56df3 in inline_memcpy /mnt/disk1/wangqiannan/amory/doris/be/src/glibc-compatibility/memcpy/memcpy_x86_64.cpp:202
    #2 0x564829a56eb0 in memcpy /mnt/disk1/wangqiannan/amory/doris/be/src/glibc-compatibility/memcpy/memcpy_x86_64.cpp:219
    #3 0x564831035cb7 in doris::vectorized::ColumnVector<unsigned char>::insert_range_from(doris::vectorized::IColumn const&, unsigned long, unsigned long) /mnt/disk1/wangqiannan/amory/doris/be/src/vec/columns/column_vector.cpp:362
    #4 0x564830fe099c in doris::vectorized::ColumnNullable::insert_range_from(doris::vectorized::IColumn const&, unsigned long, unsigned long) /mnt/disk1/wangqiannan/amory/doris/be/src/vec/columns/column_nullable.cpp:247
    #5 0x564831222231 in doris::vectorized::ColumnMap::insert_range_from(doris::vectorized::IColumn const&, unsigned long, unsigned long) /mnt/disk1/wangqiannan/amory/doris/be/src/vec/columns/column_map.cpp:255
    #6 0x564830fe0a1e in doris::vectorized::ColumnNullable::insert_range_from(doris::vectorized::IColumn const&, unsigned long, unsigned long) /mnt/disk1/wangqiannan/amory/doris/be/src/vec/columns/column_nullable.cpp:248
    #7 0x56483e62a9d7 in doris::vectorized::VerticalMergeIteratorContext::copy_rows(doris::vectorized::Block*, unsigned long) /mnt/disk1/wangqiannan/amory/doris/be/src/vec/olap/vertical_merge_iterator.cpp:267
    #8 0x56483e631406 in doris::vectorized::VerticalMaskMergeIterator::next_batch(doris::vectorized::Block*) /mnt/disk1/wangqiannan/amory/doris/be/src/vec/olap/vertical_merge_iterator.cpp:585
    #9 0x56483e64f082 in doris::vectorized::VerticalBlockReader::_direct_next_block(doris::vectorized::Block*, bool*) /mnt/disk1/wangqiannan/amory/doris/be/src/vec/olap/vertical_block_reader.cpp:207
    #10 0x564829d194fc in doris::vectorized::VerticalBlockReader::next_block_with_aggregation(doris::vectorized::Block*, bool*) /mnt/disk1/wangqiannan/amory/doris/be/src/vec/olap/vertical_block_reader.h:45
    #11 0x564829cfe9d3 in doris::Merger::vertical_compact_one_group(std::shared_ptr<doris::Tablet>, doris::ReaderType, std::shared_ptr<doris::TabletSchema>, bool, std::vector<unsigned int, std::allocator<unsigned int> > const&, doris::vectorized::RowSourcesBuffer*, std::vector<std::shared_ptr<doris::RowsetReader>, std::allocator<std::shared_ptr<doris::RowsetReader> > > const&, doris::RowsetWriter*, long, doris::Merger::Statistics*) /mnt/disk1/wangqiannan/amory/doris/be/src/olap/merger.cpp:216
    #12 0x564829d01cd2 in doris::Merger::vertical_merge_rowsets(std::shared_ptr<doris::Tablet>, doris::ReaderType, std::shared_ptr<doris::TabletSchema>, std::vector<std::shared_ptr<doris::RowsetReader>, std::allocator<std::shared_ptr<doris::RowsetReader> > > const&, doris::RowsetWriter*, long, doris::Merger::Statistics*) /mnt/disk1/wangqiannan/amory/doris/be/src/olap/merger.cpp:323
    #13 0x564829c9ff25 in doris::Compaction::do_compaction_impl(long) /mnt/disk1/wangqiannan/amory/doris/be/src/olap/compaction.cpp:287
    #14 0x564829c9bab5 in doris::Compaction::do_compaction(long) /mnt/disk1/wangqiannan/amory/doris/be/src/olap/compaction.cpp:74
    #15 0x56482a6ff2e0 in doris::BaseCompaction::execute_compact_impl() /mnt/disk1/wangqiannan/amory/doris/be/src/olap/base_compaction.cpp:76
    #16 0x564829c9b606 in doris::Compaction::execute_compact() /mnt/disk1/wangqiannan/amory/doris/be/src/olap/compaction.cpp:55
    #17 0x56482a7bf969 in doris::Tablet::execute_compaction(doris::CompactionType) /mnt/disk1/wangqiannan/amory/doris/be/src/olap/tablet.cpp:1652
    #18 0x564829c22097 in operator() /mnt/disk1/wangqiannan/amory/doris/be/src/olap/olap_server.cpp:653
    #19 0x564829c3177b in __invoke_impl<void, doris::StorageEngine::_submit_compaction_task(doris::TabletSharedPtr, doris::CompactionType)::<lambda()>&> /mnt/disk1/wangqiannan/tool/ldb_toolchain/include/c++/11/bits/invoke.h:61
    #20 0x564829c2e6f6 in __invoke_r<void, doris::StorageEngine::_submit_compaction_task(doris::TabletSharedPtr, doris::CompactionType)::<lambda()>&> /mnt/disk1/wangqiannan/tool/ldb_toolchain/include/c++/11/bits/invoke.h:111
    #21 0x564829c2c5ef in _M_invoke /mnt/disk1/wangqiannan/tool/ldb_toolchain/include/c++/11/bits/std_function.h:291
```

Describe your changes.
 I calculate the pre-disk offsets data  with offsets data in memory using relative offset 

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by amorynan** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.12 seconds
 stream load tsv:          442 seconds loaded 74807831229 Bytes, about 161 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          73 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230402044227_clickbench_pr_123735.html

**Comment by amorynan** [comment]
@xy720 @xiaokang please review this pr

File: be/src/vec/olap/olap_data_convertor.cpp [hunk 1]
```diff
@@ -893,28 +893,44 @@ Status OlapBlockDataConvertor::OlapColumnDataConvertorMap::convert_to_olap(
...
-            _base_row); // _offsets must start with current map offsets which coming blocks in continue pages
-    for (auto it = offsets.begin(); it < offsets.end(); ++it) {
-        _offsets.push_back(*it + _base_row);
-    }
-
     int64_t start_index = _row_pos - 1;
     int64_t end_index = _row_pos + _num_rows - 1;
-    auto start = offsets[start_index];
-    auto size = offsets[end_index] - start;
+    auto start_offset = offsets[start_index];
```
Comment (xiaokang) [inline be/src/vec/olap/olap_data_convertor.cpp:898]: use column_map->offset_at(_row_pos)

**Review by xiaokang — DISMISSED** [review]
LGTM

File: be/src/vec/olap/olap_data_convertor.cpp [hunk 2]
```diff
@@ -893,28 +893,44 @@ Status OlapBlockDataConvertor::OlapColumnDataConvertorMap::convert_to_olap(
...
+    auto size = offsets[end_index] - start_offset;
+
+    // NOTICE here are two situation:
+    // 1. Multi-SegmentWriter with different olap_convertor to convert same column_map(in memory which is from same block)
+    //   eg: Block(6 row): column_map offsets in memory: [10, 21, 33, 43, 54, 66]
+    //   After SegmentWriter1 with olap_convertor1 deal with first 3 rows:  _offsets(pre-disk)=[0, 10, 21], _base_row=33
+    //   then Maybe SegmentWriter will flush data (see BetaRowsetWriter::_add_block(max_row_add < 1))
```
Comment (xy720) [inline be/src/vec/olap/olap_data_convertor.cpp:905]:
```suggestion
    //   then SegmentWriter may flush data (see BetaRowsetWriter::_add_block(max_row_add < 1))
```

File: be/src/vec/olap/olap_data_convertor.cpp [hunk 3]
```diff
@@ -893,28 +893,44 @@ Status OlapBlockDataConvertor::OlapColumnDataConvertorMap::convert_to_olap(
...
     int64_t end_index = _row_pos + _num_rows - 1;
-    auto start = offsets[start_index];
-    auto size = offsets[end_index] - start;
+    auto start_offset = offsets[start_index];
+    auto size = offsets[end_index] - start_offset;
+
+    // NOTICE here are two situation:
+    // 1. Multi-SegmentWriter with different olap_convertor to convert same column_map(in memory which is from same block)
+    //   eg: Block(6 row): column_map offsets in memory: [10, 21, 33, 43, 54, 66]
+    //   After SegmentWriter1 with olap_convertor1 deal with first 3 rows:  _offsets(pre-disk)=[0, 10, 21], _base_row=33
```
Comment (xy720) [inline be/src/vec/olap/olap_data_convertor.cpp:904]:
```suggestion
    //   After SegmentWriter1 with olap_convertor1 deal with first 3 rows:  _offsets(pre-disk)=[0, 10, 21], _base_offset=33
```

File: be/src/vec/olap/olap_data_convertor.cpp [hunk 4]
```diff
@@ -893,28 +893,44 @@ Status OlapBlockDataConvertor::OlapColumnDataConvertorMap::convert_to_olap(
...
+    //                                                 ---------  ----
+    //                                                 |--_offsets  |--set_next_array_item_ordinal(_kv_writers[0]->get_next_rowid())
+    //   new SegmentWriter2 with olap_convertor2 deal with next map offsets [43, 54, 66]
+    //   but in disk here is new segment file offset should start with 0, so after convert:
+    //      _offsets(pre-disk)=[0, 10, 21], _base_row=33, After flush data finally in disk: [0, 10, 21, 33]
+    //2. One-SegmentWriter with olap_convertor to convertor different blocks into one page
+    //    eg: Two blocks -> block1 [10, 21, 33] and block2 [1, 3, 6]
+    //      After first convert: _offsets_1(pre-disk)=[0, 10, 21], _base_row=33, without flush, just append to page,
+    //      then deal with coming block2, after current convert:
+    //      _offsets_2=[33, 34, 36], _base_row=39
```
Comment (xy720) [inline be/src/vec/olap/olap_data_convertor.cpp:916]:
```suggestion
    //      _offsets_2=[33, 34, 36], _base_offset=39
```

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run p0

**Comment by amorynan** [comment]
run p0

**Comment by amorynan** [comment]
run arm

**Comment by amorynan** [comment]
run arm

**Comment by amorynan** [comment]
run p1

**Review by xy720 — APPROVED** [review]
LGTM

**Comment by amorynan** [comment]
@xy720  please merge. all test passed

**Pull request merged**